### ISSUE TITLE
Python expose EnumMap and InfoArray cleanly without causing a template mess

### ISSUE THREAD

**Issue by Nightinggale** [We-the-People-civ4col-mod/Mod#941]
One major limitation of EnumMap is that it's not python exposed. It would have been by now if it wasn't for the fact that it's heavily templated. We can't just add a pointer like we add pointers to vanilla (non template) classes.

We have two solutions. One is to have CyEnumMap, which has an int array, which is set by a template constructor. This makes a copy of the data and as such makes the array read only. This is fine considering #938, but it's limiting in case we do want to have write access. Also some EnumMaps can be rather big, so making a full copy each time might not be resource efficient.

The other solution would be to make an enum for CyEnumMap types and then store an enum instance and a reference to an EnumMap. This would then require a switch-case in each function and a template constructor.

```cpp
class CyEnumMap
{
public:
    enum CyEnumMapTypes
    {
        CyEnumMap_YieldTypes_int,
    };

    int get(int IIndex) const;

protected:
    m_type CyEnumMapTypes;
    union
    {
        EnumMap<YieldTypes, int>& EM_YieldTypes_int;
    }
};

int CyEnumMap::get(int iIndex) const
{
    switch(CyEnumMapTypes)
    {
        case CyEnumMap_YieldTypes_int: return EM_YieldTypes_int.get((YieldTypes)iIndex);
    }
}

```

This will work, but the amount of code needed for each makes it tempting to script generate CyEnumMap.

It's also possible to do both, as in a "self configuring" read only CyEnumMap and a more complex one, which allows writing. 

-----

Same consideration for InfoArray, except they are always read only. Also 99% of the time they are significantly shorter, so less data to copy.

InfoArrayBase is legacy code, which contains the actual data while InfoArray inherits it and is a wrapper layer to provide type safe access. CyInfoArray is the only code outside of IntoArray, which makes use of InfoArrayBase. Without InfoArrayBase we can change InfoArray to be "real" template, which moves a lot of checks from runtime to compile time and each template of InfoArray can have a struct of the data, which is then stored in a vector. Not only does this show much better in the debugger, it also unlocks the ability to return a const vector of all the data.

In short cleaning up CyInfoArray can make C++ InfoArrays much more useful and powerful.

**Comment by Nightinggale** [comment]
I think I might have been over thinking this. I don't think we need a non-template class as a bridge to expose a template class. We can expose a specialized instance of a template class instead. Nobody said we can't expose EnumArray<YieldTypes, bool> and EnumMap<UnitTypes, int> as two different exposed entities.

```cpp
python::class_<CvUnitInfo>("CvUnitInfo")
```
That is what we need to expose. A C++ class inside <> and then a name for python. Vanilla has a tendency to name those two the same, but I don't think that is a requirement. It makes for a good logical design, so in general that is a good choice.

We can make a standard class interface with those two being T and const char* and then put that declaration in a template function. We should then be able to add more using just a single line, like
```cpp
EnumMapInterface(EnumMap<YieldTypes, int>, "EnumMapYieldInt");
```
I see no reason why this shouldn't work. Apart from using wildcards early in the compile process, a template class and a non-template class should be the same.

If we are a bit clever, we place python interface and normal implementation in the same file. That way if a specific template combo is missing from getting an interface, then no such instance will be present in the cpp file, which in turn will make all the functions for that template combo not compile. If that template combo is used elsewhere, then we will get a C++ compile time error due to missing constructor. In other words it should be impossible to return CyInfoArray or CyEnumMap without also having that specific one python exposed.

**Comment by johan-buret** [comment]
The system looks interesting, but for now it only works with hardcoded values, as Python is initialized before XML is loaded.

**Comment by Nightinggale** [comment]
> The system looks interesting, but for now it only works with hardcoded values, as Python is initialized before XML is loaded.

Python initializes after xml lengths and types have been loaded, but before any other xml data is loaded. This means the data required to expose template EnumMap and InfoArray will be known at python expose time. In fact only the enum types will be required to be known when exposing the interface and those are compile time variable types.
